src/classes/CW_AudienceCSVProcessHelper.cls

```diff
@@ -0,0 +1,406 @@
+		CampaignID=CsvStatus.Marketing_Campaign__c;
+	}
+	
+	private String getFldValue(List<String> flds , integer Idx){
+		if(flds.Size()<=Idx)
+			return null;
+		String sVal=flds.get(Idx);
+		if(sVal==null)
+			return null;
+		if(!String.isNotBlank(sVal.trim()))
+			return null;
+		return sVal.trim();
+	}
+	
+	private String ToUpper(String sVal){
+		if(sVal==null) return '';
+		return sVal.ToUpperCase();
+	}
+	     // CSV contains Account.Palyer_Id__c and CW_Audience_Group__r.Name,FreeBet ID, Ammount, currency
+    public void processGroupMemberRewards(List<List<string>> csvLines)
+    {
+        map<String,CW_CSV_Stage_AudienceMembers__c> MembrRewardsMap = new map<String,CW_CSV_Stage_AudienceMembers__c>();
+        map<string,id> GroupsIdmap=new map<string,id>();
+        map<string,id> PlayerIdmap=new map<string,id>();
+        map<string,id> FreeBetIdmap=new map<string,id>();
+        for(List<String> flds : csvLines)
+        {
+            CW_CSV_Stage_AudienceMembers__c CsvRec = new CW_CSV_Stage_AudienceMembers__c(CSV_Import_Status__c =csvStatus.Id);
+            CsvRec.Group_Name__c =getFldValue(flds,ConstantUtil.CSV_GROUP_NAME_COLUMN);
+            CsvRec.PlayerId__c   =getFldValue(flds,ConstantUtil.CSV_ACCOUNT_PLAYER_ID_COLUMN);
+				
+				if(CsvStatus.Reward_Catalogue__c!=null){// If is null there no need to load this column.
+					CsvRec.FreeBetId__c  =getFldValue(flds,ConstantUtil.CSV_FREEBET_ID_COLUMN);
+					if(getFldValue(flds,ConstantUtil.CSV_AMOUNT_COLUMN)!=null )
+						CsvRec.Amount__c     =Double.valueOf(getFldValue(flds,ConstantUtil.CSV_AMOUNT_COLUMN));
+					CsvRec.Currency__c   =getFldValue(flds,ConstantUtil.CSV_CURRENCY_COLUMN);
+				}
+				if(CsvRec.Group_Name__c==null || CsvRec.PlayerId__c==null) //Not upload if names a null
+					continue;
+            if(!String.isNotBlank(CsvRec.Group_Name__c))//Emty group name skip record
+            	continue;
+            if(!String.isNotBlank(CsvRec.PlayerId__c))//Emty group name skip record
+            	continue;
+            CsvRec.Name=CsvRec.Group_Name__c+'|'+CsvRec.PlayerId__c+'|'+CsvRec.FreeBetId__c;
+            MembrRewardsMap.Put(CsvRec.Name,csvRec); //Uniqu records -> Group_PlayerId
+            GroupsIdmap.Put(ToUpper(CsvRec.Group_Name__c),null);
+            PlayerIdmap.Put(ToUpper(CsvRec.PlayerId__c),null);
+            FreeBetIdmap.Put(ToUpper(CsvRec.FreeBetId__c),null);
+        }
+        //Find Players
+        for(Contact cntRec: [Select ID,Account_Name__c FROM Contact WHERE Account_Name__c=:PlayerIdmap.KeySet() ]){
+        		//Contact.Account_Name__c
+        		PlayerIdmap.Put(ToUpper(cntRec.Account_Name__c),cntRec.ID);
+        }
+        //Find groups
+        for(CW_Audience_Group__c grpRec: [Select ID,Name FROM CW_Audience_Group__c WHERE Name=:GroupsIdmap.KeySet() AND Campaign__c =:CsvStatus.Marketing_Campaign__c]){
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.